Yandex: add id system

chernodub opened the pull request:
<!--
Thank you for your pull request! 

Please title your pull request like this: 'Module: Change', eg 'Fraggles Bid Adapter: support fragglerock'

Please make sure this PR is scoped to one change or you may be asked to resubmit. 
 
Please make sure any added or changed code includes tests with greater than 80% code coverage. 

See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.

For any user facing change, submit a link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [x] New ID system

PR to the docs repo: https://github.com/prebid/prebid.github.io/pull/5268


## Description of change

Added yandex id system that gets the id provided by [Yandex Metrica tag](https://metrica.yandex.com/list) that is optionally installed by the publishers. In case the id is not set by the tag, it will be set from the yandexIdSystem sumbodule.

chernodub commented:
Hi @ChrisHuie! Our team would really appreciate it if you (or anybody from the Prebid Team) could look into this PR. If anything needs to be updated/explained — I'm here. Thanks!

ChrisHuie commented:
@chernodub looks like there is a CodeQL issue around insecure randomness and linting issues with single quotes. Can you please update?

chernodub commented:
Hi @ChrisHuie! Thanks for the opportunity to improve the code. I've added the changes you recommended 🙏

1. Added the tests.
2. Submitted the docs PR: https://github.com/prebid/prebid.github.io/pull/5268
3. Fixed the insecure randomness issue by using the Web Crypto API. I had to fall back to `Math.random` in environments that do not support the crypto module. As far as I see, several other modules also use the `Math.random` in a similar context, so probably this won't cause any issues. But I'd be happy to add the updates if you have any thoughts on this.
4. Fixed the codestyle.

Please let me know if anything else needs to be added/improved. Thanks!

bretg reviewed (commented):
the docs PR https://github.com/prebid/prebid.github.io/pull/5268 suggests there's a bidders parameter ... not seeing this in the code?

chernodub commented on modules/yandexIdSystem.js:
```diff
@@ -0,0 +1,137 @@
+  },
+  eids: {
+    [YANDEX_ID_KEY]: {
+      source: BIDDER_CODE,
```
Hi @bretg, thanks! 

1. I changed the eid key to domain name. Is it a convention? I wonder what's the reason. We at Yandex have different region-specific domains (e.g. `.com`, `.kz`, etc) so I'm not sure if it's a good idea hardcoding the TLD here.
2. Considering `bidders`. As far as I understood, this is not a userland parameter, but an option provided by Prebid SDK to control which bidders will receive the eid — [here is the link to the docs](https://docs.prebid.org/dev-docs/modules/userId.html#how-it-works:~:text=Array%20of%20Strings-,An%20array%20of%20bidder%20codes%20to%20which%20this%20user%20ID%20may%20be%20sent,-.). Are you sure we need to handle that manually in the module code? I tested this end-to-end I see it working fine without any adjustments on module's side. I also see the related references of `config.bidders` in the core code.

bretg replied:
the eids source key is what DSPs look for when they code for a given ID. It's possible that you only care about your own bidder so the value of the eids source isn't as important, but yes, the convention is that it's a domain.

we're good on the bidders part. thanks.

chernodub commented:
Hi @ChrisHuie! All the updates are done. I'd really appreciate it if you could look into this once more. Thank you!

chernodub commented:
Hi! Could anybody look into this PR, please? This seems to be lost/forgotten. Thanks!

cc @ChrisHuie @bretg

patmmccann commented on modules/yandexIdSystem.js:
```diff
@@ -0,0 +1,138 @@
+    if (this._cookieStorage.cookiesAreEnabled()) {
+      const expires = new Date(Date.now() + USER_ID_COOKIE_EXP_MS).toString();
+
+      this._cookieStorage.setCookie(USER_ID_KEY, userId, expires);
```
The parent module can handle setting the id accorring to publisher config, the submodules only needs to gather it

patmmccann replied:
See #10710

chernodub replied:
Hi @patmmccann, thanks for the recommendation! **Is it required by Prebid for us to maintain different types of storage and allow custom keys?** Can we just fall back to "do nothing"+"log error" when `storage` option is customized for our submodule?

I'm asking this because this submodule is intended to be a thin bridge between our Yandex Metrika publisher tag and our bidder adapter. And we want it to be kept as simple as possible.

We intentionally want only to use the cookie and constant `USER_ID_KEY`.

chernodub commented:
Hi @patmmccann! Huge thanks for pointing out the way to use the parent module for managing the cookie. I've updated the code, now it looks way simpler! I've also updated the tests and [the docs](https://github.com/prebid/prebid.github.io/pull/5268/commits/a680fc0db535f396a8429dd7feef7e6855efab43)
